rust/src/module_expressions.rs

```diff
@@ -0,0 +1,191 @@
+use crate::errors::{GrimpError, GrimpResult};
+use itertools::Itertools;
+use lazy_static::lazy_static;
+use regex::Regex;
+use std::fmt::Display;
+use std::str::FromStr;
+
+lazy_static! {
+    static ref MODULE_EXPRESSION_PATTERN: Regex =
+        Regex::new(r"^(\w+|\*{1,2})(\.(\w+|\*{1,2}))*$").unwrap();
+}
+
+/// A module expression is used to refer to sets of modules.
+///
+/// - `*` stands in for a module name, without including subpackages.
+/// - `**` includes subpackages too.
+#[derive(Debug, Clone)]
+pub struct ModuleExpression {
+    expression: String,
+    pattern: Regex,
+}
+
+impl Display for ModuleExpression {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "{}", self.expression)
+    }
+}
+
+impl AsRef<str> for ModuleExpression {
+    fn as_ref(&self) -> &str {
+        &self.expression
+    }
+}
+
+impl FromStr for ModuleExpression {
+    type Err = GrimpError;
+
+    fn from_str(expression: &str) -> Result<Self, Self::Err> {
+        if !MODULE_EXPRESSION_PATTERN.is_match(expression) {
+            return Err(GrimpError::InvalidModuleExpression(expression.to_owned()));
+        }
+
+        for (part, next_part) in expression.split(".").tuple_windows() {
+            match (part, next_part) {
+                ("*", "**") | ("**", "*") | ("**", "**") => {
+                    return Err(GrimpError::InvalidModuleExpression(expression.to_owned()))
+                }
+                _ => {}
+            }
+        }
+
+        Ok(Self {
+            expression: expression.to_owned(),
+            pattern: Self::create_pattern(expression)?,
+        })
+    }
+}
+
+impl ModuleExpression {
+    pub fn is_match(&self, module_name: &str) -> bool {
+        self.pattern.is_match(module_name)
+    }
+
+    fn create_pattern(expression: &str) -> GrimpResult<Regex> {
+        let mut pattern_parts = vec![];
+
+        for part in expression.split(".") {
+            if part == "*" {
+                pattern_parts.push(part.replace("*", r"[^\.]+"))
+            } else if part == "**" {
+                pattern_parts.push(part.replace("**", r"[^\.]+(?:\.[^\.]+)*?"))
+            } else {
+                pattern_parts.push(part.to_owned());
+            }
+        }
+
+        Ok(Regex::new(&(r"^".to_owned() + &pattern_parts.join(r".") + r"$")).unwrap())
```

Comment:
Should be `Ok(Regex::new(&(r"^".to_owned() + &pattern_parts.join(r"\.") + r"$")).unwrap())` i.e. `\.` instead of `.`. Sorry, this one slipped past me 🙈